dwave/optimization/include/dwave-optimization/nodes/lambda.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#pragma once
+
+#include <vector>
+
+#include "dwave-optimization/array.hpp"
+#include "dwave-optimization/graph.hpp"
+#include "dwave-optimization/nodes/constants.hpp"
+
+namespace dwave::optimization {
+
+class NaryReduceNode : public ArrayOutputMixin<ArrayNode> {
+ public:
+    // Runtime constructor that can be used from Cython/Python
+    NaryReduceNode(Graph&& expression, const std::vector<InputNode*>& inputs,
```

Comment:
I wonder if it makes more sense to not have the user provide inputs etc explicitly, but rather to inspect the `expression`. Maybe `set_objetive()` sets the output? And we can traverse the nodes looking for `InputNode`

Reply:
I think inputs should be provided explicitly because they need to be mapped to the operands/predecessors. Otherwise we are relying on the implicit ordering of when the nodes are added to the expression model, and it becomes less clear which one is the "special" input that takes the previous value.

But I still find this interface confusing... was thinking about taking a map between inputs/operands as well.

Reply:
At the Python level this could perhaps be addressed by not adding an `input()` method, but having a subclass of `Model` that requires `num_inputs` at construction-time. Then we rely on the order? We might have to treat inputs as decision variables in that they get topologically ordered immediately in that case... Needs thought.

Reply:
I like that. The `ExpressionModel` or whatever (subclass of `Model`) could disallow adding decisions.

Perhaps we could also somehow pull a list of supported nodes from C++ and then disallow adding unsupported symbols to `ExpressionModel`. This will go a long way to prevent constructing unsupported expressions, as I don't think we will have any supported nodes that can have non-scalar output given all scalar inputs.

Reply:
And I'll note I expect the standard flow when creating expressions is just to create a list of `Input`s anyway and use that. So `exp = ExpressionModel(7)`, `exp.inputs[0] + exp.inputs[1]` feels natural to me.